src/raftstore/store/worker/apply.rs

```diff
@@ -1105,6 +1151,270 @@ impl ApplyDelegate {
+                self.tag,
+                first_index,
+                exist_first_index,
+                source_region.get_id(),
+                e
+            );
+        });
+        entries.extend_from_slice(merge.get_entries());
+        entries
+    }
+
+    fn catch_up_log_for_merge(
+        &mut self,
+        ctx: &mut ApplyContext,
+        merge: &CommitMergeRequest,
+        exist_region: &Region,
+    ) {
+        let source_region = merge.get_source();
+        let apply_state_key = keys::apply_state_key(source_region.get_id());
+        let apply_state: RaftApplyState = match self.engine.get_msg_cf(CF_RAFT, &apply_state_key) {
+            Ok(Some(s)) => s,
+            e => panic!(
+                "{} failed to get apply state of {:?}: {:?}",
+                self.tag, source_region, e
+            ),
+        };
+        let apply_index = apply_state.get_applied_index();
+        if apply_index >= merge.get_commit() {
+            if source_region.get_region_epoch() != exist_region.get_region_epoch() {
+                panic!(
+                    "{} source region {:?} not match exist region {:?}",
+                    self.tag, source_region, exist_region
+                );
+            }
+            return;
+        }
+
+        let entries = self.load_entries_for_merge(merge, apply_index);
+        if entries.is_empty() {
+            return;
+        }
+        let stash = ctx.stash(self);
+        let mut delegate = match ctx.delegates.get_mut(&source_region.get_id()) {
+            None => panic!("{} source region {:?} not exist", self.tag, source_region),
+            Some(e) => e.take().unwrap_or_else(|| {
+                panic!(
+                    "{} unexpected circle dependency of region {:?}",
+                    self.tag, source_region
+                )
+            }),
+        };
+        delegate.handle_raft_committed_entries(ctx, entries);
+        *ctx.delegates.get_mut(&source_region.get_id()).unwrap() = Some(delegate);
+        ctx.apply_res.last_mut().unwrap().merged = true;
+        ctx.restore_stash(stash);
+    }
+
+    fn exec_commit_merge(
+        &mut self,
+        ctx: &mut ApplyContext,
+        req: &AdminRequest,
+    ) -> Result<(AdminResponse, Option<ExecResult>)> {
+        PEER_ADMIN_CMD_COUNTER_VEC
+            .with_label_values(&["commit_merge", "all"])
+            .inc();
+
+        let merge = req.get_commit_merge();
+        let source_region = merge.get_source();
+        let region_state_key = keys::region_state_key(source_region.get_id());
+        let state: RegionLocalState = match self.engine.get_msg_cf(CF_RAFT, &region_state_key) {
+            Ok(Some(s)) => s,
+            e => panic!(
+                "{} failed to get regions state of {:?}: {:?}",
+                self.tag, source_region, e
+            ),
+        };
+        match state.get_state() {
+            PeerState::Normal | PeerState::Merging => {}
```

Comment:
can the state be Normal here?

Reply:
Yes, the follower may not apply `PrepareMerge` yet.